app/scripts/modules/core/src/presentation/forms/hooks/useSaveRestoreMutuallyExclusiveFields.hook.ts

```diff
@@ -0,0 +1,69 @@
+ * "pizza", this hook saves a copy of the user's currently entered "sandwich" data and then clears the "sandwich" fields.
+ * If the user then switches back to "sandwich", the hook restores the "sandwich" data and saves/clears the "pizza" fields.
+ *
+ * Note: this hook does not explicitly attempt to manage default values for form fields
+ *
+ * @param formik formik props (typically from the Formik render function)
+ * @param currentFieldSetKey the current form mode (e.g., the value of the selected radio button)
+ * @param mutuallyExclusiveFieldSets an object with:
+ *        - keys: one for each form mode
+ *        - values: lodash paths to the form fields exclusively owned by that form mode
+ *
+ * Example:
+ *
+ * // Restores previous "pizza" form data ('toppings', 'crust', 'sauce') when the form mode switches back to 'pizza'
+ * // Restores previous "sandwich" form data ('bread', 'meat', 'cheese') when the form mode switches back to 'sandwich'
+ *
+ * useSaveRestoreMutuallyExclusiveFields(formik, pizzaOrSandwich, {
+ *   { pizza: ["toppings", "crust", "sauce" ] },
+ *   { sandwich: ["bread", "meat", "cheese" ] },
+ * });
+ *
+ * <FormikFormField name="pizzaOrSandwich" input={props => <SelectInput {...props} options={['pizza', 'sandwich']}>} />
+ *
+ * { formik.values.pizzaOrSandwich === 'pizza' && (<>
+ *   <FormikFormField name="toppings" input={props => <SelectInput {...props} options={toppingsOptions}>} />
+ *   <FormikFormField name="crust" input={props => <SelectInput {...props} options={crustOptions}>} />
+ *   <FormikFormField name="sauce" input={props => <SelectInput {...props} options={sauceOptions}>} />
+ * </>)}
+ *
+ * { formik.values.pizzaOrSandwich === 'sandwich' && (<>
+ *   <FormikFormField name="bread" input={props => <SelectInput {...props} options={breadOptions}>} />
+ *   <FormikFormField name="meat" input={props => <SelectInput {...props} options={meatOptions}>} />
+ *   <FormikFormField name="cheese" input={props => <SelectInput {...props} options={cheeseOptions}>} />
+ * </>)}
+ */
+export function useSaveRestoreMutuallyExclusiveFields(
+  formik: FormikProps<any>,
+  currentFieldSetKey: string,
+  mutuallyExclusiveFieldSets: { [fieldSetId: string]: string[] },
+) {
+  const previousFormMode = usePrevious(currentFieldSetKey);
+  const [savedData, setSavedData] = useState({} as { [path: string]: any });
+
+  // Whenever the user switches the form mode source, clear out the
+  // previous form mode's exclusive field values and restore the new mode's values.
+  useEffect(() => {
+    if (!!previousFormMode && currentFieldSetKey !== previousFormMode) {
+      const fieldsToSave = mutuallyExclusiveFieldSets[previousFormMode] ?? [];
+      const fieldsToRestore = mutuallyExclusiveFieldSets[currentFieldSetKey] ?? [];
+
+      const savedDataCopy = { ...savedData };
+      fieldsToSave.forEach(path => (savedDataCopy[path] = get(formik.values, path)));
+      setSavedData(savedDataCopy);
```

Comment:
I generally think of using the following form of data structure to store the state for multiple modes. I'm curious on why you chose to save all the keys in the same level. Is it 'cos when someone uses it accidentally to store non-mutually exclusive fields, we want to restore the latest value for those fields?

```js
const savedData: {[fieldSetId: string]: {[path: string]: any}};
```

Reply:
great point!  I'll update the PR with your suggestion.

Reply:
With your suggestion, both pizza and sandwich can have their own copy of a `cheese` property.